packages/alfa-rules/src/sia-r69/diagnostic/contrast.ts

```diff
@@ -0,0 +1,125 @@
+import { Diagnostic } from "@siteimprove/alfa-act";
+import { RGB } from "@siteimprove/alfa-css";
+import { Equatable } from "@siteimprove/alfa-equatable";
+import { Serializable } from "@siteimprove/alfa-json";
+
+import * as json from "@siteimprove/alfa-json";
+
+export class Contrast extends Diagnostic {
+  public static of(
+    message: string,
+    threshold: number = 4.5,
+    pairings: Iterable<Contrast.Pairing> = []
+  ): Contrast {
+    return new Contrast(message, threshold, Array.from(pairings));
+  }
+
+  private readonly _threshold: number;
+  private readonly _pairings: Array<Contrast.Pairing>;
+
+  private constructor(
+    message: string,
+    threshold: number = 4.5,
+    pairings: Array<Contrast.Pairing>
+  ) {
+    super(message);
+
+    this._threshold = threshold;
+    this._pairings = pairings;
+  }
+
+  public get threshold(): number {
+    return this._threshold;
+  }
+
+  public get pairings(): Iterable<Contrast.Pairing> {
+    return this._pairings;
+  }
+
+  public equals(value: unknown): value is this {
+    return (
+      super.equals(value) &&
+      value instanceof Contrast &&
+      value._threshold === this._threshold &&
+      value._pairings.length === this._pairings.length &&
+      value._pairings.every((pairing, i) => pairing.equals(this._pairings[i]))
+    );
+  }
+
+  public toJSON(): Contrast.JSON {
+    return {
+      ...super.toJSON(),
+      threshold: this._threshold,
+      pairings: this._pairings.map((pairing) => pairing.toJSON()),
+    };
+  }
+}
+
+export namespace Contrast {
+  export interface JSON extends Diagnostic.JSON {
+    threshold: number;
+    pairings: Array<Pairing.JSON>;
+  }
+
+  export function isContrast(value: unknown): value is Contrast {
+    return value instanceof Contrast;
+  }
+
+  export class Pairing implements Equatable, Serializable {
+    public static of(
+      foreground: RGB,
+      background: RGB,
+      contrast: number
```

Comment:
Should we move the contrast computation here?
That would especially allow to write rule for enhanced contrast by re-using the computation in the shared "contrast diagnostic" instead of repeating it in the rule.

Given that contrast computation is an "absolute" external formula, I think it could make sense to delegate it here. OTOH, this gives extra responsibilities (computing contrast) to the diagnostic which is only a structured data storage and this is maybe not a good way to go 🤔

Reply:
Yeah, I'm not too keen on moving logic into the diagnostics 🤔 The contrast computation should absolutely be moved out of the current rule, but the diagnostics aren't the destination.

Reply:
Yes, I agree that keeping diagnostics as a clean "storage only" structure is a good idea to keep separation of concern. They're just the paper on which the doctor wrote the conclusion, but not really part of the actual examination…

As for contrast computation, it will probably move to `common/expectations/` once sharing is needed.